torch/csrc/api/include/torch/detail/any.h

```diff
@@ -0,0 +1,138 @@
+#include <utility>
+
+namespace torch {
+class Any;
+namespace detail {
+template <typename T>
+Any make_any(T&& value);
+} // namespace detail
+} // namespace torch
+
+namespace torch {
+/// A simplified implementation of `std::any` or `boost::any` which stores a
+/// type erased object, whose concrete value can be retrieved at runtime by
+/// checking if the `typeid()` of a requested type matches the `typeid()` of the
+/// object stored. It is simplified in that it does not handle copying, as we do
+/// not require it for our use cases. Moves are sufficient.
+class Any {
+ public:
+  /// Default construction is disallowed. Thus our invariant: an `Any` contains
+  /// an object at all times, from its construction, to its destruction.
+  Any() = delete;
+
+  /// Move construction and assignment is allowed, and follows the default
+  /// behavior of move for `std::unique_ptr`.
+  Any(Any&&) = default;
+  Any& operator=(Any&&) = default;
+
+  /// Copy is disallowed, because we don't need it.
+  Any(const Any& other) = delete;
+  Any& operator=(const Any& other) = delete;
+
+  /// Returns the value contained in the `Any` if the type passed as template
+  /// parameter matches the type of the object stored, and returns a null
+  /// pointer otherwise.
+  template <typename T>
+  T* try_get() {
+    static_assert(
+        !std::is_reference<T>::value,
+        "Any stores decayed types, you cannot cast it to a reference type");
+    static_assert(
+        !std::is_array<T>::value,
+        "Any stores decayed types, you must cast it to T* instead of T[]");
+    if (typeid(T).hash_code() == type_info().hash_code()) {
+      return &static_cast<Holder<T>&>(*content_).value;
+    }
+    return nullptr;
+  }
+
+  template <typename T>
+  T get() {
+    if (auto* value = try_get<T>()) {
+      return *value;
+    }
+    AT_ERROR(
+        "Attempted to cast Any to",
+        at::demangle(typeid(T).name()),
+        ", but its contained type is ",
+        at::demangle(type_info().name()));
+  }
+
+  /// Returns the `type_info` object of the contained value.
+  const std::type_info& type_info() const noexcept {
+    return content_->type_info;
+  }
+
+ private:
+  /// Constructs the `Any` from any type.
+  template <typename T>
+  explicit Any(T&& value)
+      : content_(
+            torch::make_unique<Holder<decay_t<T>>>(std::forward<T>(value))) {}
+
+  /// `Any` is a public type, but its construction is only allowed via this
+  /// "private" function in the `detail` namespace.
+  template <typename T>
+  friend Any detail::make_any(T&& value);
+
+  /// The static type of the object we store in the `Any`, which erases the
+  /// actual object's type, allowing us only to check the `type_info` of the
+  /// type stored in the dynamic type.
+  struct Placeholder {
+    explicit Placeholder(const std::type_info& type_info_) noexcept
+        : type_info(type_info_) {}
+    virtual ~Placeholder() = default;
+    const std::type_info& type_info;
```

Comment:
Worth noting that this is a reference because `type_info` is not copyable.

Reply:
And because `typeid()` returns this type

Reply:
isn't there a `type_index` type as well? Or does that need RTTI?